ds701_book/27-RNN.qmd

```diff
@@ -0,0 +1,640 @@
+:::
+::::
+
+## Loss Calculation for Sequential Data
+
+For each prediction in the sequence $y_t$ and corresponding true value $\hat{y}_t$  we calculate the the cross-entropy loss $\mathcal{L}_t(\hat{y}_t, y_t) = -\hat{y}_t\log(y_t)$. The total loss is the sum of all the individual losses, $\mathcal{L} = \sum_t \mathcal{L}_t$. This is illustrated in the figure below.
+
+If we were predicting continious value we would use a mean-square error loss.
+
+![](drawio/RNN-Loss.png)
+
+
+## Vanishing Gradients
+
+The key idea of RNNs is that earlier items in the sequence  influence the more recent outputs. This is highlighted in blue in the figure below.
+
+However, for longer sequences the influence can be significantly reduced. This is illustrated in red in the figure below.
+
+Due to the potential for long sequences of data, during training you are likely to encounter small gradients. As a result, RNNs are prone to vanishing gradients.
+
+![](drawio/VanishingGradient.png)
+
+## RNN Limitations
+
+- **Vanishing Gradients**:
+  - During training, gradients can become very small (vanish), making it difficult for the network to learn long-term dependencies.
+
+- **Long-Term Dependencies**:
+  - RNNs struggle to capture dependencies that are far apart in the sequence, leading to poor performance on tasks requiring long-term memory.
+
+- **Difficulty in Parallelization**:
+  - The sequential processing of data in RNNs makes it challenging to parallelize training, leading to slower training times.
+
+- **Limited Context**:
+  - Standard RNNs have a limited ability to remember context over long sequences, which can affect their performance on complex tasks.
+
+## Addressing RNN Limitations
+
+Some proposed solutions for mitigating the previous issues are
+
+- **LSTM (Long Short-Term Memory)**
+- **GRU (Gated Recurrent Units)** 
+
+These are more advanced variants of RNNs designed to address some of these limitations by improving the ability to capture long-term dependencies and mitigating gradient issues.
+
+## LSTM
+
+Below is an illustration of the LSTM architecture.
+
+![](drawio/LSTM.png)
+
+---
+
+The new components are:
+
+- The input cell state $c_{t-1}$ and output cell state $c_{t}$.
+- Yellow blocks consisting of neural network layers with sigmoid $\sigma$ or tanh activation functions.
+- Red circles indicating point-wise operations.
+
+
+## Cell State
+
+:::: {.columns}
+::: {.column width="50%"}
+
+The cell state $c_{t-1}$ is the input to the LSTM block.
+
+This value then moves through the LSTM block.
+
+It is modified by either a multiplication or addition interaction. 
+
+After these operations, the modified input state becomes $c_{t}$ and is sent to the next LSTM block.
+
+:::
+::: {.column width="50%"}
+![](drawio/CellState.png)
+:::
+::::
+
+## Forget Gate Layer
+
+:::: {.columns}
+::: {.column width="50%"}
+The forget layer computes a value 
+
+$$
+f_t = \sigma(W_f[h_{t-1}, x_t] + b_f),
+$$
+
+where $\sigma$ is the sigmoid function.
```

Comment:
Should we add a reminder of what the sigmoid function is?

Reply:
Added.